### PR TITLE
Orchestrator HTTP/gRPC health checks and health readiness probe

### PR DESCRIPTION
**Pull request by pscoro**
### Issue
https://github.ibm.com/ai-foundation/watson-fm-stack-tracker/issues/1222
OSS: https://github.com/foundation-model-stack/fms-guardrails-orchestrator/issues/163

### Changes
- Health checks supported by orchestrator for both gRPC and HTTP:
    - any client's health status can be `READY`, `NOT_READY`, or `UNKNOWN`
    - gRPC: Introduce health check service proto (using the v1 health check proto officially provided by gRPC: https://github.com/grpc/grpc/blob/master/src/proto/grpc/health/v1/health.proto, introduce corresponding health client for each grpc client.
    - HTTP: uses a minimal health contract
        - health endpoint: `/health`
        - empty `200 OK` -> `READY`
        - non-empty `200 OK` -> try parse health_status (e.g. `{ "health_status": "UNKNOWN" }`), if HTTP response body is not empty this is required. Optionally a `reason` can also be provided `{ "health_status": "NOT_READY", "response_code": "HTTP 200 OK", "reason": "some reason" }`. This was done in order to allow http endpoints to potentially use `200 OK` for any successful response even if the response is to say `NOT_READY` or `UNKNOWN`, so if the client just wants to return `READY`, a body can/should be omitted. If a 200 OK response is to be considered `READY` and the response does contains a body, it must have `"health_status": "READY"`, but if `"reason": "..."` is specified it will be ignored as this is meant as an additional field for unhealthy reporting
        - `503 SERVICE UNAVAILABLE` and all other `5xx` HTTP codes are treated as `NOT_READY` because of the implication that a `5xx` error is the fault of the server.
        - `404 NOT FOUND` and all other `4xx` HTTP codes are treated as `UNNKNOWN` because of the implication that a `4xx` error is the fault of the client, and so the most we know is that we do not know the health status and that this could likely be caused by misconfiguration/misuse rather than a server issue
- Implement readiness probe for the actual client "handlers" (generation, detection, chunker) to collect the health check results of the clients
- expose readiness probe through orchestrator and introduce a health status cache in orchestrator for clients
    - introduce `/ready` server endpoint (as separate from `/health` which is a superficial orchestrator server check)
    - `/ready` takes `?probe=true/false` query param with default `false`, when `true`, a readiness probe and update of health status cache is triggered, otherwise just read cache.
    - returns a serialization of `HealthStatusCache`, showing status for each client in detectors, chunkers, generation. 
        -  `READY` clients' statuses are simplified `"some_client_id": "READY"`, while `NOT_READY` and `UNKNOWN` statuses include more information (e.g. `{ "health_status": "UNKNOWN", "response_code": "HTTP 404 Not Found", "reason": "HTTP status client error (404 Not Found) for url (http://no.url.provided.local/): { \"message\": \"service not found\" }" }`)
    - readiness probe is triggered on orchestrator start up and (only at startup) `NOT_READY` or `UNKNOWN` statuses are considered fatal for the server.
- added unit tests mocking responses from HTTP and gRPC clients, testing handling of the different successful and unsuccessful cases.
- Included a commit that updates the OpenAPI spec, if this is wanted.

### CODE REVIEW

**Review by declark1 — COMMENTED** [review]
Looks pretty good to me. I left some comments inline, mostly nits.

I noticed the calls to the underlying clients in the `HealthProbe` impl aren't performed concurrently, which is probably not an issue for the generation clients as I anticipate those to actually be single "router" clients. If we have a lot of chunkers and detectors, we may want to consider doing so, depending on the probe frequency.

File: src/health.rs [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "overall health status: {}\n{}",
+            self.health_status, self.services
+        )
+    }
+}
+
+impl Serialize for HealthCheckResult {
```
[the remark this answers is not in the document]
Comment (pscoro) [reply src/health.rs:248]: I didnt want to clutter a response that contains mostly `READY` responses with full objects for each client and wanted a shorthand string status:

What the response should look like:
```json
{
    "health_status": "NOT_READY",
    "services": {
        "generation": {
            "model_id_1": "READY",
            "model_id_2": "READY",
            "model_id_3": {
                "health_status": "NOT_READY",
                "response_code": "HTTP 503 Service Unavailable",
                "reason": "HTTP status server error (503 Service Unavailable) for url (http://no.url.provided.local/): { \"message\": \"some error message\" }"
            },
            ...
        },
        "detectors": { ... },
        "chunkers": { ... },
    }
}
```

**Comment by pscoro** [comment]
Left an explanation about one of the comments and resolved nits from all the others

**Comment by pscoro** [comment]
Update to start up behaviour:

Based on discussion regarding requirements for start up validation, I've updated the PR so that the client health statuses are not considered part of a validation process on orchestrator start-up. The orchestrator still performs the client health probe on start-up, but only so that it can have the results ready in cache (TODO: do we want a flag for this?). The client health probe does not provide an overall status, as it now only relays the health/readiness of clients and makes no inference about the orchestrator itself. The client services are not dependencies of the orchestrator. Failing client health checks will not impede the orchestrator's liveliness during start-up or after any other re-probing. The client health probe is now exposed by `/info` instead of `/ready`, a `?probe=true` can still be provided. Below is an example of the logging output on orchestrator startup for a config that contains clients that fail to pass health checks:

image

**Comment by pscoro** [comment]
Issue arising from a conflict with the recently merged introduction of integration tests by @mdevino: https://github.com/foundation-model-stack/fms-guardrails-orchestrator/pull/188

Summary of problem and discussion so far: 

- Background info - a merge conflict for health probe PR caused by recent introductory integration test PR causes the questions about mocking w.r.t. integration tests to need addressing

- Problematic detail - we already use some mocking for unit tests with e.g. #[cfg_attr(test, faux::create, derive(Default))] , however the built-in build configuration condition test is purposely not enabled by rust for integration tests (tests in tests/)
Why not? - Taking a step back, the point of integration tests is to test the integration of all modules of the code together, as they would actually be used. Mocking is generally a discouraged practice for integration tests as you can not test that all the parts of a system are working together when some of the parts are mocked. Rust adheres to this premise and thats why conditional build configurations for integration tests are not enabled by the test feature.

- Short term fix - when I found this issue on Friday I made a temporary solution where a custom cargo feature "mock" acting in place of test is made required by the integration test instead, and the build configuration attributes in the code were modified to #[cfg_attr(any(test, feature = "mock"), faux::create, derive(Default))]. Building off the previous point though, perhaps we want to re-evaluate our approach to integration testing to avoid mocking parts of the system

- Speculation about a real fix - IMO I think a more correct approach to integration testing should not mock any orchestrator code, instead it can mock the client services. Maybe something like an orchestrator config file for very simple localhost instances of client services (just simple scripts mocking the client service behaviour) that return dummy responses without any additional logic

- What about functional tests? - The main difference in purpose I see between our functional tests and the proposal for integration tests is that functional tests are for validating the business logic while integration tests are testing that the components of the system interact as expected. As such, functional tests are black box tests that validate the response contains correct logic given a request, while an integration test is a (white box?) test that ensure a request is handled and responded to as expected without caring about the contents of the dummy request/responses used

As it stands, neither @mdevino or I have a clear direction about what the correct approach is and believe this should be discussed further with the team.

**Comment by pscoro** [comment]
cc @declark1 I resolved the failing chunker health check issue (caikit expects an empty string to be passed for the `service_id` of a `HealthCheckRequest` and specifying the chunker service id was causing the failure) and addressed the other discovered integration test issue. This PR is now ready for final review

File: src/health.rs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        matches!(self.health_status, HealthStatus::Ready)
+    }
+
+    pub fn is_not_ready(&self) -> bool {
+        matches!(self.health_status, HealthStatus::NotReady)
+    }
+
+    pub fn is_unknown(&self) -> bool {
+        matches!(self.health_status, HealthStatus::Unknown)
+    }
```
Comment (declark1) [inline src/health.rs:112]: None of these methods are actually used, so they should be dropped

File: src/health.rs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+                    response.status
+                );
+                Some(format!(
+                    "Unexpected gRPC health check response status: {}",
+                    response.status
+                ))
+            }
+        }
+    }
+}
```
Comment (declark1) [inline src/health.rs:140]: These descriptions feel redundant, aren't the status strings such as UNKNOWN, NOT_SERVING, etc.  self-explanatory?

File: src/clients.rs [hunk 4]
```diff
@@ -125,20 +128,123 @@ impl From<tonic::Status> for Error {
...
     }
 
     pub fn base_url(&self) -> &Url {
         &self.base_url
     }
+
+    /// This is sectioned off to allow for testing.
+    pub(super) async fn http_response_to_health_check_result(
+        res: Result<Response, reqwest::Error>,
+    ) -> HealthCheckResult {
```
Comment (declark1) [inline src/clients.rs:170]: All of the logic in this method feels unnecessary IMO. I think we can simply consider a non-200 status code on a health endpoint to be unhealthy? 

Also, I think `HealthStatus::NotReady` should be renamed to `HealthStatus::Unhealthy`, as we are checking health, not readiness, and `HealthStatus::Unknown` should be dropped as a service is either healthy or it isn't, why introduce the concept of "unknown" here? Unknown essentially means unhealthy in this context.

**Review by declark1 — CHANGES_REQUESTED** [review]
I left a couple of comments inline as I think some of this can be simplified and unused methods should be dropped. Also, I suggest that we rename `HealthStatus::NotReady` to `HealthStatus::Unhealthy` (we are checking health, not readiness), and I think `HealthStatus::Unknown` should be dropped as a service is either healthy or it isn't; unknown essentially means unhealthy in this context.

File: src/health.rs (hunk 3 above)
Comment (pscoro) [reply src/health.rs:140]: Regardless of whether we use `HEALTHY | UNHEALTHY | UNKNOWN` or just `HEALTHY | UNHEALTHY`, we can preserve the original gRPC status in non-healthy cases as part of the reason message for debugging, where gRPC health status is of `READY | NOT_READY | SERVICE_UNKNOWN | UNKNOWN`. I think this is something nice to preserve, although if these are too wordy maybe simpler messages like just `from gRPC serving_status: SERVICE_UNKNOWN`, omitting the redundant `Service's health is currently unknown`?

**Review by declark1 — APPROVED** [review]
Per discussion, we will merge this as-is and later open a separate PR to make health checks configurable/optional.
